### PR TITLE
Implement `REDEF` to allow redefining `EQUS` string equates

### LINKED ISSUE #677: [Feature request] A way to explicitly re-assign an EQU or EQUS constant
Unlike `SET` constants, `EQU` and `EQUS` ones can't be reassigned, which is convenient to prevent changing one by accident that you didn't know already existed. However, it means that intentionally reassigning one has to do it step-by-step with `PURGE`:

```
TMP EQUS "{FOO}, more"
    PURGE FOO
FOO EQUS "{TMP}"
    PURGE TMP
```

I'd appreciate a keyword to intentionally reassign a name. Like:

```
FOO RE_EQUS "{FOO}, more"

FOO UPDATE EQUS "{FOO}, more"

FOO EQUS FORCE "{FOO}, more"
```

(Haven't thought of a good keyword for it though.)

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #677

### CODE REVIEW

**Comment by Rangi42** [comment]
I implemented this as `REDEF str EQUS "value"` rather than `str REDEF EQUS "value"` for symmetry with `PURGE sym`.

**Comment by ISSOtm** [comment]
Also because `str REDEF` would be impossible to use.

File: src/asm/rgbasm.5 [hunk 1]
```diff
@@ -1002,8 +1002,21 @@ pusha EQUS "push af\[rs]npush bc\[rs]npush de\[rs]npush hl\[rs]n"
 Note that colons
 .Ql \&:
 following the name are not allowed.
+.Pp
 String equates can't be exported or imported.
 .Pp
+String equates, like
+.Ic EQU
+constants, cannot be redefined. However, the
```
Comment (ISSOtm) [inline src/asm/rgbasm.5:1010]: Sentence-ending periods must end at EOL.

File: src/asm/symbol.c [hunk 2]
```diff
@@ -222,6 +222,21 @@ static void fullSymbolName(char *output, size_t outputSize,
...
 }
 
+static void assignStringSymbol(struct Symbol *sym, char const *value)
+{
+	size_t len = strlen(value);
+	char *string = malloc(len + 1);
+
+	if (string == NULL)
+		fatalerror("No memory for string equate: %s\n", strerror(errno));
+	strcpy(string, value);
```
Comment (ISSOtm) [inline src/asm/symbol.c:232]: Isn't all that just `strdup`?

File: src/asm/symbol.c [hunk 3]
```diff
@@ -390,17 +405,25 @@ struct Symbol *sym_AddEqu(char const *symName, int32_t value)
...
+
+	if (!sym) {
+		sym = createsymbol(symName);
+	} else if (sym->type != SYM_EQUS) {
+		error("'%s' already defined as non-EQUS at ", symName);
+		dumpFilename(sym);
+		putc('\n', stderr);
+	}
+
+	assignStringSymbol(sym, value);
```
Comment (ISSOtm) [inline src/asm/symbol.c:426]: Careful, this leaks the old string.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/symbol.c (hunk 2 above)
Comment (Rangi42) [reply src/asm/symbol.c:232]: Now that you mention it, yes. I'd just moved that code out of `sym_AddString`.

File: src/asm/symbol.c (hunk 3 above)
Comment (Rangi42) [reply src/asm/symbol.c:426]: I think `sym_Purge` is leaking strings too. It calls `hash_RemoveElement` and `free(symbol)`, but not `free(symbol->macro)` for `SYM_MACRO` or `SYM_EQUS`.

Comment (Rangi42) [reply src/asm/symbol.c:426]: Fixed for `sym_RedefString`.

File: src/asm/symbol.c (hunk 2 above)
Comment (Rangi42) [reply src/asm/symbol.c]: Simplified to `strdup`.

File: src/asm/symbol.c (hunk 3 above)
Comment (ISSOtm) [reply src/asm/symbol.c:426]: That ought to be fixed too, then; whether in this PR or separately, I'll leave to you.

Comment (Rangi42) [reply src/asm/symbol.c:426]: Actually, that causes equs-purge and macro-purge to fail, since they purge themselves.

Comment (ISSOtm) [reply src/asm/symbol.c:426]: If they are self-referenced? Geh. Please add a `FIXME` comment instead, then...

Comment (Rangi42) [reply src/asm/symbol.c:431]: `FIXME` added.

Comment (ISSOtm) [reply src/asm/symbol.c:431]: This code path doesn't have it, though. Other than that, LGTM: I'll merge once that's done.

**Review by ISSOtm — APPROVED** [review]

**Comment by Rangi42** [comment]
> This code path doesn't have it, though.

On that path it's either a new symbol or not previously an `EQUS`, so there's nothing to free.

**Comment by ISSOtm** [comment]
Right, my bad, there's a `free` just above. That said, a `REDEF` that redefines itself would be a UAF, wouldn't it?

**Comment by Rangi42** [comment]
I think it's not quite a UAF; the new value gets used instead. Which *does* give unexpected results.

```
foo equs "\nredef foo equs \" printv 42\"\n printv $100\n printt \"\\n\"\n"
 foo
 foo
 printt "\n"
```

Without the `free(sym->macro)`, this prints:

```
$100
$2A
```

With it freed:

```
ERROR: test.asm(2):
    Unknown character 0x00
while expanding symbol "foo"
ERROR: test.asm(2):
    Unknown character 0x00
while expanding symbol "foo"
ERROR: test.asm(2):
    Unknown character 0x00
while expanding symbol "foo"
ERROR: test.asm(2):
    Unknown character 0x00
while expanding symbol "foo"
ERROR: test.asm(2):
    syntax error
while expanding symbol "foo"
error: Assembly aborted (5 errors)!
```

How about just not freeing it and adding the same `FIXME` comment there? Since any code that uses `REDEF` would previously have used `PURGE` and `EQUS`, and had the same leak problem.

**Comment by ISSOtm** [comment]
Definitely.

**Comment by Rangi42** [comment]
Done; the memory leak is noted for both.

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**